x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/CachingServiceAccountsTokenStore.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc.service;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.cache.CacheBuilder;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.ListenableFuture;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.security.authc.support.Hasher;
+import org.elasticsearch.xpack.security.support.CacheInvalidatorRegistry;
+
+import java.util.Collection;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public abstract class CachingServiceAccountsTokenStore implements ServiceAccountsTokenStore, CacheInvalidatorRegistry.CacheInvalidator {
```

Comment:
I wonder if this would be better using composition rather than inheritance.
Then we can have 
```
new CachingTokenStore(new CompositeTokenStore( new FileTokenStore(), new IndexTokenStore() ) )
```
And have a single cache across all stores.

Reply:
Having a single cache for all stores is something I tried to avoid. For example, the `invalidateAll` will invalidate all entries even when only one of the store needs it. Also if the same key exists in both stores, in a bad case, they could evict each other and cause cache thrashing. We could solve all these by possibly prefix the key with the store type. But I wonder what is the better tradeoff?

Reply:
@tvernum Rest of the PR is updated to address your comment. I didn't touch this part though. Let me know if my above comment makes sense. Thanks!